### PR TITLE
Add local:// scheme support for the App jar

### PR DESCRIPTION
**Pull request by skonto**
## What changes were proposed in this pull request?

- Adds support for local:// scheme like in k8s case for image based deployments where the jar is already in the image. Mimics the k8s approach.

## How was this patch tested?

dispatcher image with the patch, use it to start DC/OS Spark service: 
skonto/spark-local-dispatcher:test

Test image with my application jar located at the root folder:
skonto/spark-local:test

Dockerfile for that image.

From mesosphere/spark:2.3.0-2.2.1-2-hadoop-2.6
COPY spark-examples_2.11-2.2.1.jar /
WORKDIR /opt/spark/dist

Tests:

The following work as expected:

* local normal example
```
dcos spark run --submit-args="--conf spark.mesos.appJar.local.resolution.mode=container --conf spark.executor.memory=1g --conf spark.mesos.executor.docker.image=skonto/spark-local:test
 --conf spark.executor.cores=2 --conf spark.cores.max=8
 --class org.apache.spark.examples.SparkPi local:///spark-examples_2.11-2.2.1.jar"
```

* make sure the flag does not affect other uris
```
dcos spark run --submit-args="--conf spark.mesos.appJar.local.resolution.mode=container --conf spark.executor.memory=1g  --conf spark.executor.cores=2 --conf spark.cores.max=8
 --class org.apache.spark.examples.SparkPi https://s3-eu-west-1.amazonaws.com/fdp-stavros-test/spark-examples_2.11-2.1.1.jar"
```

* normal example no local
```
dcos spark run --submit-args="--conf spark.executor.memory=1g  --conf spark.executor.cores=2 --conf spark.cores.max=8
 --class org.apache.spark.examples.SparkPi https://s3-eu-west-1.amazonaws.com/fdp-stavros-test/spark-examples_2.11-2.1.1.jar"

```

The following fails

 * uses local with no setting, default is host.
```
dcos spark run --submit-args="--conf spark.executor.memory=1g --conf spark.mesos.executor.docker.image=skonto/spark-local:test
  --conf spark.executor.cores=2 --conf spark.cores.max=8
  --class org.apache.spark.examples.SparkPi local:///spark-examples_2.11-2.2.1.jar"
```

image

### CODE REVIEW

**Comment by skonto** [comment]
@susanxhuynh pls review.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 1]
```diff
@@ -430,17 +428,41 @@ private[spark] class MesosClusterScheduler(
...
+        true
+      } else {
+        if (value != "host") {
+          logWarning("Unknown spark.mesos.appJar.local.resolution.mode value, using host.")
+        }
+        false
+      }
+      case None => false
+    }
+    isLocalJar && isContainerLocal
```
[the remark this answers is not in the document]
Comment (skonto) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: The idea is that we want to keep also the default option of resolving jars on host.

File: resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala [hunk 2]
```diff
@@ -430,17 +428,41 @@ private[spark] class MesosClusterScheduler(
     envBuilder.build()
   }
 
+
+  private def isContainerLocalAppJar(desc: MesosDriverDescription): Boolean = {
+
+    val isLocalJar = desc.jarUrl.startsWith("local://")
+    val isContainerLocal = desc.conf.getOption("spark.mesos.appJar.local.resolution.mode") match {
```
[the remark this answers is not in the document]
Comment (skonto) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: ok

[the remark this answers is not in the document]
Comment (skonto) [reply resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala]: Even better:
exists {
      case "container" => true
      case "host" => false
      case other =>
        logWarning(s"Unknown spark.mesos.appJar.local.resolution.mode $other, using host.")
        false
    }

**Comment by skonto** [comment]
@susanxhuynh I am going to push this upstream.

**Comment by skonto** [comment]
https://github.com/apache/spark/pull/21378

**Comment by susanxhuynh** [comment]
Passing CI now.

**Comment by skonto** [comment]
closing this.

**Pull request closed without merging**